ui/src/app/views/notebook-resource-card.tsx

```diff
@@ -0,0 +1,188 @@
+import {workspacesApi} from 'app/services/swagger-fetch-clients';
+import colors from 'app/styles/colors';
+import {navigateByUrl} from 'app/utils/navigation';
+import {ResourceType} from 'app/utils/resourceActions';
+import {CopyNotebookModal} from 'app/views/copy-notebook-modal';
+import {RenameModal} from 'app/views/rename-modal';
+import {Action, ResourceCardTemplate} from 'app/views/resource-card-template';
+import {RecentResource} from 'generated/fetch';
+import * as React from 'react';
+import * as fp from 'lodash';
+import {withErrorModal} from "app/views/with-error-modal";
+import {withConfirmDeleteModal} from "app/views/with-confirm-delete-modal";
+
+interface Props {
+  resourceCard: RecentResource; // Destructure this into used parameters only
+  onUpdate: Function; // TODO eric: this could use a better name
+  onDuplicateResource: Function;
+  showErrorModal: Function;
+  showConfirmDeleteModal: Function;
+}
+
+interface State {
+  showCopyNotebookModal: boolean;
+  renaming: boolean;
+}
+
+export const NotebookResourceCard = fp.flow(
+  withErrorModal(),
+  withConfirmDeleteModal(),
+)(class extends React.Component<Props, State> {
+
+    constructor(props: Props) {
+      super(props);
+      this.state = {
+        showCopyNotebookModal: false,
+        renaming: false
+      };
+    }
+
+    get actions(): Action[] {
+      return [
+        {
+          displayName: 'Rename',
+          onClick: () => {
+            this.setState({renaming: true});
+          },
+        },
+        {
+          displayName: 'Duplicate',
+          onClick: (resourceCardFns) => this.duplicateNotebook(),
+        },
+        {
+          displayName: 'Copy to another Workspace',
+          onClick: () => this.setState({showCopyNotebookModal: true}),
+        },
+        {
+          displayName: 'Delete',
+          onClick: () => {
+            this.props.showConfirmDeleteModal(this.displayName, 'Notebook',
+              (onComplete) => this.deleteNotebook(onComplete));
```

Comment:
Instead of having an onComplete callback passed through here, could you have the receiveDelete function be typed to always return a Promise? That may be a slightly-cleaner control flow (but only marginally cleaner, so not a huge deal if you leave as-is).

Reply:
tried it out and it looks good. first time actually creating my own promises